File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/FlowRegister.java
package com.sequenceiq.cloudbreak.core.flow2;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;
import org.springframework.stereotype.Component;

@Component
public class FlowRegister {
    private Map<String, Pair<Flow, String>> runningFlows = new ConcurrentHashMap<>();

    public void put(Flow flow, String chainFlowId) {
        runningFlows.put(flow.getFlowId(), new ImmutablePair<>(flow, chainFlowId));
    }

    public Flow get(String flowId) {
        Pair<Flow, String> p = runningFlows.get(flowId);
        return p != null ? p.getLeft() : null;
    }

    public String getFlowChainId(String flowId) {
        Pair<Flow, String> p = runningFlows.get(flowId);
        return p != null ? p.getRight() : null;
    }

    public Flow remove(String flowId) {
        Pair<Flow, String> pair = runningFlows.remove(flowId);
        return pair == null ? null : pair.getLeft();
    }
}

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/FlowRegister.kt
package com.sequenceiq.cloudbreak.core.flow2

import java.util.concurrent.ConcurrentHashMap

import org.apache.commons.lang3.tuple.ImmutablePair
import org.apache.commons.lang3.tuple.Pair
import org.springframework.stereotype.Component

@Component
class FlowRegister {
    private val runningFlows = ConcurrentHashMap<String, Pair<Flow, String>>()

    fun put(flow: Flow, chainFlowId: String) {
        runningFlows.put(flow.flowId, ImmutablePair(flow, chainFlowId))
    }

    operator fun get(flowId: String): Flow {
        return runningFlows[flowId].getLeft()
    }

    fun getFlowChainId(flowId: String): String {
        return runningFlows[flowId].getRight()
    }

    fun remove(flowId: String): Flow? {
        val pair = runningFlows.remove(flowId)
        return pair?.left
    }
}
